Sources/Signal.swift

```diff
@@ -2015,177 +1876,352 @@ private enum ThrottleWhileState<Value> {
+			values[position] = value
+
+			if isInitialized {
+				action(values)
+			}
+
+			return false
+		}
+
+		mutating func complete(at position: Int) -> Bool {
+			completionCount += 1
+			return completionCount == values.count
+		}
+
+		init(count: Int, action: @escaping (ContiguousArray<Any>) -> Void) {
+			values = ContiguousArray(repeating: Placeholder(), count: count)
+			completionCount = 0
+			_isInitialized = false
+			self.action = action
+		}
+	}
+
+	private struct ZipStrategy: SignalAggregateStrategy {
+		private var values: ContiguousArray<[Any]>
+		private var isCompleted: ContiguousArray<Bool>
+		private let action: (ContiguousArray<Any>) -> Void
+
+		mutating func update(_ value: Any, at position: Int) -> Bool {
+			values[position].append(value)
+
+			if values.reduce(true, { $0 && !$1.isEmpty }) {
+				var buffer = ContiguousArray<Any>()
+				buffer.reserveCapacity(values.count)
+
+				for index in values.startIndex ..< values.endIndex {
+					buffer.append(values[index].removeFirst())
+				}
+
+				action(buffer)
+
+				if Swift.zip(values, isCompleted).contains(where: { $0.isEmpty && $1 }) {
+					return true
+				}
+			}
+
+			return false
+		}
+
+		mutating func complete(at position: Int) -> Bool {
+			isCompleted[position] = true
+
+			// `zip` completes when all signals has completed, or any of the signals
+			// has completed without any buffered value.
+			return Swift.zip(values, isCompleted).contains { $0.isEmpty && $1 }
+				|| isCompleted.reduce(true) { $0 && $1 }
+		}
+
+		init(count: Int, action: @escaping (ContiguousArray<Any>) -> Void) {
+			values = ContiguousArray(repeating: [], count: count)
+			isCompleted = ContiguousArray(repeating: false, count: count)
+			self.action = action
+		}
+	}
+
+	private final class AggregateBuilder<Strategy: SignalAggregateStrategy> {
+		private var startHandlers: [(Int, CompositeDisposable, Atomic<Strategy>, @escaping (Event<Never, Error>) -> Void) -> Void]?
+
+		init() {
+			self.startHandlers = []
+		}
+
+		@discardableResult
+		func add<U>(_ signal: Signal<U, Error>) -> Self {
+			precondition(startHandlers != nil, "Signals cannot be registered after the state has started.")
+
+			startHandlers!.append { index, disposables, strategy, action in
+				disposables += signal.observe { event in
+					switch event {
+					case let .value(value):
+						let shouldComplete = strategy.modify {
+							return $0.update(value, at: index)
+						}
+
+						if shouldComplete {
+							action(.completed)
+						}
+
+					case .completed:
+						let shouldComplete = strategy.modify {
+							return $0.complete(at: index)
+						}
+
+						if shouldComplete {
+							action(.completed)
+						}
+
+					case .interrupted:
+						action(.interrupted)
+
+					case let .failed(error):
+						action(.failed(error))
+					}
+				}
+			}
+
+			return self
+		}
+
+		private func make<Composite>(count: Int, _ transform: @escaping (ContiguousArray<Any>) -> Composite) -> Signal<Composite, Error> {
```

Comment:
Yeah. 😞 But until the type system is powerful enough, I don't mind some internal type erasure.